src/librustc/middle/ty.rs

```diff
@@ -2301,6 +2289,177 @@ impl<'a, 'tcx> ParameterEnvironment<'a, 'tcx> {
     }
 }
 
+pub struct DatatypeDef<'tcx> {
+    pub def_id: DefId,
+    pub variants: Vec<VariantDef<'tcx>>,
+}
+
+pub struct VariantDef<'tcx> {
+    pub id: DefId,
+    pub name: Name,
+    pub disr_val: Disr,
+    pub fields: Vec<FieldTy<'tcx>>,
+    pub vis: Visibility,
+}
+
+
+pub struct FieldTy<'tcx> {
+    // The DefId of the field itself.
+    pub id: DefId,
+    pub name: Name,
+    pub vis: ast::Visibility,
+    // The "parent" of the field, the DefId of the struct or enum this field comes from
+    pub origin: DefId,
+    // `Cell` is used here to to allow the construction of a DatatypeDef that contains types
+    // that reference this type. For example:
+    //
+    //     struct Foo(Bar);
+    //     struct Bar(Option<Box<Foo>>);
+    //
+    // Making the `ty_struct` for `Foo` requires the type of `Bar`, but making the `type`
+    // of `Bar` requires the type of `Foo`. The `Cell` is initially set to `ty_err` and
+    // the real type is filled in later
+    ty: Cell<Ty<'tcx>>
+}
+
+impl<'tcx> DatatypeDef<'tcx> {
+
+    pub fn is_tuple_variant(&self, variant_idx: usize) -> bool {
+        if self.variants.len() <= variant_idx { return false; }
+
+        self.variants[variant_idx].is_tuple_variant()
+    }
+
+    pub fn get_variant(&'tcx self, def: DefId) -> Option<&'tcx VariantDef<'tcx>> {
+        self.variants.iter().find(|v| v.id == def)
+    }
+
+    pub fn is_univariant(&self) -> bool {
+        self.variants.len() == 1
+    }
+
+    pub fn is_struct_def(&self) -> bool {
+        self.is_univariant() && self.def_id == self.variants[0].id
+    }
+
+    pub fn is_c_like(&self) -> bool {
+        if self.variants.len() == 0 {
+            return false;
+        }
+
+        self.variants.iter().all(|v| v.fields.len() == 0)
+    }
+}
+
+impl<'tcx> VariantDef<'tcx> {
+    pub fn subst_fields<'a>(&'a self, cx: &'a ctxt<'tcx>,
+                            substs: &'tcx Substs<'tcx>) -> FieldTyIter<'a, 'tcx> {
+        FieldTyIter {
+            cx: cx,
+            substs: substs,
+            iter: self.fields.iter()
+        }
+    }
+
+    pub fn is_tuple_variant(&self) -> bool {
+        !self.fields.is_empty() &&
+            self.fields.iter().all(|f| f.is_unnamed())
+    }
+}
+
+impl<'tcx> FieldTy<'tcx> {
+    pub fn new(tcx: &ctxt<'tcx>,
+               id: DefId, name: Name, vis: ast::Visibility, origin: DefId) -> FieldTy<'tcx> {
+        FieldTy {
+            id: id,
+            name: name,
+            vis: vis,
+            origin: origin,
+            ty: Cell::new(tcx.types.err),
+        }
+    }
+
+    pub fn is_unnamed(&self) -> bool {
+        self.name == token::special_names::unnamed_field
+    }
+
+    pub fn set_ty(&self, ty: Ty<'tcx>) {
+        if let ty_err = self.ty().sty {
+            self.ty.set(ty);
+        } else {
+            panic!("FieldTy::set_ty: Field type already set");
+        }
+    }
+
+    pub fn ty(&self) -> Ty<'tcx> {
+        self.ty.get()
+    }
```

Comment:
The current setup, iiuc, initially stores a ty_err and then updates it to the actual type. This seems like it will hide bugs where we read the type of a field before it is fully initialized. I would prefer to use `Cell<Option<Ty<'tcx>>>` and initially set it to `None`. Then `ty` can call `self.ty.get().unwrap()`, which will give the same (stronger) guarantees we get today. Similarly `set_ty` can check that the type is set exactly once. (There is no space inefficiency because `Ty<'tcx>` is a non-nullable-pointer.)